### DISCUSSION TITLE
Some of the labels in the glTF sample model are hard to read in Filament v1.9.14 or later

### DISCUSSION THREAD

**Discussion by cx20** [google/filament#3727] [Q&A] [answered]
I have recently noticed that some of the glTF samples are having difficulty reading the labels.
This seems to happen in Filament `v1.9.14` and later. Is this a matter of usage?
The environment will be the WebGL version of Filament.

|Filament v1.9.13|Filament v1.9.14|
|:---:|:---:|
|![image](https://user-images.githubusercontent.com/1833979/113180789-1d533080-928c-11eb-9284-0be20f64e164.png)|![image](https://user-images.githubusercontent.com/1833979/113180803-2217e480-928c-11eb-84ae-c44997dc91a6.png)|

The following are some examples of samples where it is difficult to identify the labels.
[Filament v1.9.18 + TextureEncodingTest.gltf](https://cx20.github.io/gltf-test/examples/filament/index.html?category=tutorialModels&model=TextureEncodingTest&scale=0.2&type=glTF)
[Filament v1.9.18 + TextureLinearInterpolationTest.gltf](https://cx20.github.io/gltf-test/examples/filament/index.html?category=tutorialModels&model=TextureLinearInterpolationTest&scale=0.5&type=glTF)
[Filament v1.9.18 + ClearCoatTest.gltf](https://cx20.github.io/gltf-test/examples/filament/index.html?category=tutorialModels&model=ClearCoatTest&scale=0.3&type=glTF)
[Filament v1.9.18 + TransmissionTest.gltf](https://cx20.github.io/gltf-test/examples/filament/index.html?category=tutorialModels&model=TransmissionTest&scale=3&type=glTF)

**Comment by prideout** [comment, marked as answer]
I don't think we changed anything with texture filtering so this might be due to one of our post-processing effects.  You can try doing `view.setPostProcessingEnabled(false)` to see if that helps.

**Reply by romainguy** [reply]
I can't think of anything we changed in .14 wrt post processing.

**Reply by prideout** [reply]
@cx20 are you saying this did *not* occur in .13?

**Reply by romainguy** [reply]
I know what's going on I think.

It's the change I made to make emissive surfaces bloom. It's easier to see what's happening using gltf_viewer.

@prideout we should revert my change, I think there's a Khronos discussion going on to allow emissive values > 1.0.

**Reply by romainguy** [reply]
Uploading a fix

**Reply by cx20** [reply]
@prideout Below is a sample of Filament `v1.9.13`.
[Filament v1.9.13 + TransmissionTest.gltf](https://rawcdn.githack.com/cx20/gltf-test/8eacae1471a083e5ebdea04b8e4046faf3802e20/examples/filament/index.html?category=tutorialModels&model=TransmissionTest&scale=3&type=glTF)

**Reply by romainguy** [reply]
With the [fix I uploaded](https://github.com/google/filament/pull/3728):

<img width="1136" alt="Screen Shot 2021-03-31 at 10 07 25 AM" src="https://user-images.githubusercontent.com/869684/113183454-e7209d00-9208-11eb-89b5-2ea12df232c1.png">

**Reply by cx20** [reply]
@romainguy Thank you for the quick fix.
Could it be the same problem that the following T-Rex models are too bright?

Model : https://rawcdn.githack.com/BabylonJS/Exporters/9bc140006be149687be045f60b4a25cdb45ce4fc/Maya/Samples/glTF%202.0/T-Rex/trex_running.gltf

|Library|result|
|:-----:|:--------:|
|[Filament v1.9.13 + trex_running.gltf](https://rawcdn.githack.com/cx20/gltf-test/8eacae1471a083e5ebdea04b8e4046faf3802e20/examples/filament/index.html?url=https://rawcdn.githack.com/BabylonJS/Exporters/9bc140006be149687be045f60b4a25cdb45ce4fc/Maya/Samples/glTF%202.0/T-Rex/trex_running.gltf&scale=0.003)|![image](https://user-images.githubusercontent.com/1833979/113184846-a9675700-9290-11eb-8aab-d5a325371e38.png)|
|[Filament v1.9.18 + trex_running.gltf](https://cx20.github.io/gltf-test/examples/filament/index.html?url=https://rawcdn.githack.com/BabylonJS/Exporters/9bc140006be149687be045f60b4a25cdb45ce4fc/Maya/Samples/glTF%202.0/T-Rex/trex_running.gltf&scale=0.003)|![image](https://user-images.githubusercontent.com/1833979/113184994-d74c9b80-9290-11eb-8039-c686db13e58a.png)|

**Reply by romainguy** [reply]
Yes indeed. It looks like some of the samples use emissive as a substitute for unlit.

**Reply by prideout** [reply]
Hilarious, this is the emissive skin for the dinosaur:

https://rawcdn.githack.com/BabylonJS/Exporters/9bc140006be149687be045f60b4a25cdb45ce4fc/Maya/Samples/glTF%202.0/T-Rex/mtl_trex_emissive.jpg

I wonder if the artist did this, or if the Maya exporter does strange things.

**Reply by cx20** [reply]
@romainguy Thank you for letting me know.

@prideout For reference, here is what the original Maya model looks like.
https://github.com/BabylonJS/Exporters/tree/master/Maya/Samples/sourceModels/T-Rex

It seems that they have recently done the Export again, but still the emissive textures have not changed.
https://github.com/BabylonJS/Exporters/tree/pandaGaume-update-sample/Maya/Samples/glTF%202.0/T-Rex

Unfortunately, I'm not familiar with Maya, so I don't know if the problem is with the original model or with the Exporter.

**Reply by cx20** [reply]
Hmm, I tried the latest version of Filament built by Github Action, but the T-Rex skinning animation looked weird.

|Filament v1.9.14|Filament v1.9.19rc|
|:--------:|:-------------:|
|![image](https://user-images.githubusercontent.com/1833979/113180803-2217e480-928c-11eb-84ae-c44997dc91a6.png)|![image](https://user-images.githubusercontent.com/1833979/113189348-12050280-9296-11eb-8b90-9138013b6837.png)|

|Filament v1.9.18|Filament v1.9.19rc|
|:-------------:|:-------------:|
|![image](https://user-images.githubusercontent.com/1833979/113184994-d74c9b80-9290-11eb-8039-c686db13e58a.png)|![image](https://user-images.githubusercontent.com/1833979/113189774-88a20000-9296-11eb-838f-7109a2a6f674.png)|

**Reply by prideout** [reply]
Wow I will take a look.  Might be a morphing issue.  Thank you so much for catching these issues @cx20, you are a big help to us.

**Reply by romainguy** [reply]
Hmm it works fine for me with tip of tree, on macOS.

**Reply by cx20** [reply]
For reference, I used the following WebGL version of the library for testing. I tried it on Chrome and Firefox on Windows and it seemed to have similar issues.
https://github.com/google/filament/actions/runs/705760580

**Reply by cx20** [reply]
I have uploaded a sample to another branch of gltf-test. The URLs you can actually try are as follows.
I checked it on Chrome on my MacBook Air 2019 and it seemed to have similar skinning issues.
[Filament v1.9.19rc + T-Rex result](https://rawcdn.githack.com/cx20/gltf-test/5fd868730212f3dcfbb4c9c18d4a71978c1e54b7/examples/filament/index.html?url=https://rawcdn.githack.com/BabylonJS/Exporters/9bc140006be149687be045f60b4a25cdb45ce4fc/Maya/Samples/glTF%202.0/T-Rex/trex_running.gltf&scale=0.003)

**Reply by romainguy** [reply]
The issue reproduces for me in Chrome as well, but not in the native viewer.

**Reply by prideout** [reply]
The dinosaur looks ok on my machine (2019 MacBook Pro 16") in Chrome 89 on main, using a hacked version of helmet.html (I really need to add drag-and-drop support...)

Here's video.

https://user-images.githubusercontent.com/1288904/113357471-ebc47e80-92f8-11eb-85f2-54bb2131aaf7.mov

**Reply by romainguy** [reply]
Does [this](https://rawcdn.githack.com/cx20/gltf-test/5fd868730212f3dcfbb4c9c18d4a71978c1e54b7/examples/filament/index.html?url=https://rawcdn.githack.com/BabylonJS/Exporters/9bc140006be149687be045f60b4a25cdb45ce4fc/Maya/Samples/glTF%202.0/T-Rex/trex_running.gltf&scale=0.003) work for you? It breaks on Chrome 89, MacBook Pro 13" 2020 for me.

**Reply by prideout** [reply]
Ah thanks Romain, yes I can repro there.  For now I'd like to assume this is either a client issue or an already-fixed issue, since I can't repro using our latest sample code.

**Reply by cx20** [reply]
Hmm, I tried again based on the [official Filament sample](https://google.github.io/filament/webgl/helmet.html), but it seems to have skinning issues.

I tried the following libraries)
https://github.com/google/filament/actions/runs/705760580
https://github.com/google/filament/actions/runs/709933814

The environment is Windows 10 + Chrome 89.

**Reply by cx20** [reply]
Hmm, I've tried Filament `v1.9.19` but I don't think the skinning issue has been resolved.
https://cx20.github.io/gltf-test/examples/filament/index.html?category=sampleModels&model=BrainStem&scale=1&type=glTF
![image](https://user-images.githubusercontent.com/1833979/113640160-cc8d6e80-96b5-11eb-843d-5cf0ffa019ca.png)

**Reply by romainguy** [reply]
This example is broken for me in Chrome as well. It works in gltf_viewer though.

**Reply by prideout** [reply]
I suspect this isn't a Filament issue since I can't reproduce with our own sample code.  e.g. here is the dino with our new <filament-viewer> web component:

https://prideout.net/models/dino.html

**Reply by cx20** [reply]
@prideout I have confirmed that the skinning problem does not occur if I try a local copy of Filament from your web site.
https://prideout.net/models/filament.js
https://prideout.net/models/filament.wasm

However, I have a skinning problem when I use the latest version of the following module. Perhaps there is something different about it.
https://github.com/google/filament/releases/download/v1.9.19/filament-v1.9.19-web.tgz

**Reply by cx20** [reply]
I compared the sizes of the libraries and found the following differences.
Since there seem to be many differences at the source level, a simple comparison cannot be made, but it can be said that they are not identical versions.

|File Name |prideout.net|filament-v1.9.19|
|:------------|-----------:|---------------:|
|filament.js  |      265 KB|          293 KB|
|filament.wasm|    3,756 KB|        3,726 KB|

I tried copying [dino.html](https://prideout.net/models/dino.html) locally and changing the library to Filament `v1.9.19`, but I got the following error.
```javascript
filament-viewer.js:259 
Uncaught TypeError: Filament.multiplyMatrices is not a function
    at HTMLElement._updateAsset (filament-viewer.js:259)
    at HTMLElement._renderFrame (filament-viewer.js:279)
    _updateAsset @ filament-viewer.js:259
    _renderFrame @ filament-viewer.js:279
    requestAnimationFrame (async)
           :
filament-viewer.js:235 
Uncaught (in promise) TypeError: Filament.fitIntoUnitCube is not a function
    at filament-viewer.js:235
```

**Reply by prideout** [reply]
Thanks!  The versions of emscripten used in by our automated builds does not match what I use locally (or what we recommend in BUILDING) so I've fixed that with https://github.com/google/filament/pull/3763.  This will definitely fix the skinning issue that we see only in our released builds because it includes a fix for glVertexAttribIPointer (https://github.com/emscripten-core/emscripten/pull/11742)

The more recent error that you see is expected since that particular API was added after the 1.9.19 release.

Thanks again for bringing these issues to our attention.

**Reply by cx20** [reply]
I tested it with the latest build results of Github Action and confirmed that it was okay.
https://github.com/google/filament/actions/runs/722937842

|Filament v1.9.18|Filament v1.9.20rc|
|:-------------:|:-------------:|
|![image](https://user-images.githubusercontent.com/1833979/113184994-d74c9b80-9290-11eb-8039-c686db13e58a.png)|![image](https://user-images.githubusercontent.com/1833979/113739315-a231c480-973a-11eb-8eaf-dab856f222df.png)|

I have a concern with another model, but I think that's a different issue, so I'd like to close this Q&A.

**Reply by cx20** [reply]
@prideout I have tried Filament `v1.9.20` locally and it seems that the skinning problem is recurring. Can you confirm it?

https://github.com/google/filament/releases/download/v1.9.20/filament-v1.9.20-web.tgz

|File Name |prideout.net|filament-v1.9.19|filament-v1.9.20|
|:------------|-----------:|---------------:|---------------:|
|filament.js  |      265 KB|          293 KB|          294 KB|
|filament.wasm|    3,756 KB|        3,726 KB|        3,735 KB|

![image](https://user-images.githubusercontent.com/1833979/114493074-cc144b00-9c54-11eb-97b2-6861bd3a04dd.png)


It seems that the release build of Filament `v1.9.20` uses emsdk `1.39.19` instead of `2.0.12`.

https://github.com/google/filament/runs/2326770923?check_suite_focus=true#step:3:27
```
++ curl -L https://github.com/emscripten-core/emsdk/archive/1.39.19.zip
```

/cc @bejado @romainguy

**Reply by prideout** [reply]
Yes, I can repro -- it appears that our release builds are still using the old version of emscripten, even though I updated `ci-common.sh`.  This is strange.

**Reply by prideout** [reply]
Since I updated the script only 8 days ago, it probably didn't get applied to the 1.9.20 release.

I think the 1.9.21 release will be fine.  :)

**Reply by cx20** [reply]
I would like to check again when the next version is released.

**Reply by cx20** [reply]
I have confirmed that the 1.9.21 release is fine.
